{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"machine_shape":"hm","authorship_tag":"ABX9TyNKyA9C6ZkSNgFmZUiV7Ump"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Nv0CeAK4hw-L","executionInfo":{"status":"ok","timestamp":1753988613413,"user_tz":420,"elapsed":984181,"user":{"displayName":"Jiarui Cui","userId":"15689554921577522760"}},"outputId":"0ea89538-7993-4506-83b8-effe2712f5e0"},"outputs":[{"output_type":"stream","name":"stdout","text":["--2025-07-31 18:47:09--  https://egg2.wustl.edu/roadmap/data/byFileType/signal/consolidated/macs2signal/foldChange/E116-H3K4me3.fc.signal.bigwig\n","Resolving egg2.wustl.edu (egg2.wustl.edu)... 128.252.187.85\n","Connecting to egg2.wustl.edu (egg2.wustl.edu)|128.252.187.85|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 479312792 (457M)\n","Saving to: â€˜E116-H3K4me3.fc.signal.bigwigâ€™\n","\n","E116-H3K4me3.fc.sig 100%[===================>] 457.11M  1.08MB/s    in 6m 59s  \n","\n","2025-07-31 18:54:08 (1.09 MB/s) - â€˜E116-H3K4me3.fc.signal.bigwigâ€™ saved [479312792/479312792]\n","\n","--2025-07-31 18:54:08--  https://egg2.wustl.edu/roadmap/data/byFileType/signal/consolidated/macs2signal/foldChange/E116-H3K27ac.fc.signal.bigwig\n","Resolving egg2.wustl.edu (egg2.wustl.edu)... 128.252.187.85\n","Connecting to egg2.wustl.edu (egg2.wustl.edu)|128.252.187.85|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 293665780 (280M)\n","Saving to: â€˜E116-H3K27ac.fc.signal.bigwigâ€™\n","\n","E116-H3K27ac.fc.sig 100%[===================>] 280.06M  1.20MB/s    in 3m 54s  \n","\n","2025-07-31 18:58:03 (1.20 MB/s) - â€˜E116-H3K27ac.fc.signal.bigwigâ€™ saved [293665780/293665780]\n","\n","--2025-07-31 18:58:03--  https://egg2.wustl.edu/roadmap/data/byFileType/signal/consolidated/macs2signal/foldChange/E116-H3K36me3.fc.signal.bigwig\n","Resolving egg2.wustl.edu (egg2.wustl.edu)... 128.252.187.85\n","Connecting to egg2.wustl.edu (egg2.wustl.edu)|128.252.187.85|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 393506893 (375M)\n","Saving to: â€˜E116-H3K36me3.fc.signal.bigwigâ€™\n","\n","E116-H3K36me3.fc.si 100%[===================>] 375.28M  1.18MB/s    in 5m 21s  \n","\n","2025-07-31 19:03:24 (1.17 MB/s) - â€˜E116-H3K36me3.fc.signal.bigwigâ€™ saved [393506893/393506893]\n","\n","--2025-07-31 19:03:24--  https://egg2.wustl.edu/roadmap/data/byDataType/rna/expression/57epigenomes.RPKM.pc.gz\n","Resolving egg2.wustl.edu (egg2.wustl.edu)... 128.252.187.85\n","Connecting to egg2.wustl.edu (egg2.wustl.edu)|128.252.187.85|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 2677594 (2.6M) [application/x-gzip]\n","Saving to: â€˜57epigenomes.RPKM.pc.gzâ€™\n","\n","57epigenomes.RPKM.p 100%[===================>]   2.55M  1.09MB/s    in 2.3s    \n","\n","2025-07-31 19:03:27 (1.09 MB/s) - â€˜57epigenomes.RPKM.pc.gzâ€™ saved [2677594/2677594]\n","\n","--2025-07-31 19:03:27--  https://egg2.wustl.edu/roadmap/data/byDataType/rna/expression/EG.name.txt\n","Resolving egg2.wustl.edu (egg2.wustl.edu)... 128.252.187.85\n","Connecting to egg2.wustl.edu (egg2.wustl.edu)|128.252.187.85|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 1560 (1.5K) [text/plain]\n","Saving to: â€˜EG.name.txtâ€™\n","\n","EG.name.txt         100%[===================>]   1.52K  --.-KB/s    in 0s      \n","\n","2025-07-31 19:03:28 (1013 MB/s) - â€˜EG.name.txtâ€™ saved [1560/1560]\n","\n","--2025-07-31 19:03:28--  https://egg2.wustl.edu/roadmap/data/byDataType/rna/expression/Ensembl_v65.Gencode_v10.ENSG.gene_info\n","Resolving egg2.wustl.edu (egg2.wustl.edu)... 128.252.187.85\n","Connecting to egg2.wustl.edu (egg2.wustl.edu)|128.252.187.85|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 5130488 (4.9M)\n","Saving to: â€˜Ensembl_v65.Gencode_v10.ENSG.gene_infoâ€™\n","\n","Ensembl_v65.Gencode 100%[===================>]   4.89M  1.13MB/s    in 4.3s    \n","\n","2025-07-31 19:03:33 (1.13 MB/s) - â€˜Ensembl_v65.Gencode_v10.ENSG.gene_infoâ€™ saved [5130488/5130488]\n","\n"]}],"source":["!mkdir -p data/roadmap && cd data/roadmap\n","\n","# Histone modification bigWig files for cell type E116 (GM12878)\n","!wget https://egg2.wustl.edu/roadmap/data/byFileType/signal/consolidated/macs2signal/foldChange/E116-H3K4me3.fc.signal.bigwig\n","!wget https://egg2.wustl.edu/roadmap/data/byFileType/signal/consolidated/macs2signal/foldChange/E116-H3K27ac.fc.signal.bigwig\n","!wget https://egg2.wustl.edu/roadmap/data/byFileType/signal/consolidated/macs2signal/foldChange/E116-H3K36me3.fc.signal.bigwig\n","\n","# Gene expression matrix and metadata\n","!wget https://egg2.wustl.edu/roadmap/data/byDataType/rna/expression/57epigenomes.RPKM.pc.gz\n","!wget https://egg2.wustl.edu/roadmap/data/byDataType/rna/expression/EG.name.txt\n","!wget https://egg2.wustl.edu/roadmap/data/byDataType/rna/expression/Ensembl_v65.Gencode_v10.ENSG.gene_info\n","\n","# Unzip the expression file\n","!gunzip -f 57epigenomes.RPKM.pc.gz\n","\n"]},{"cell_type":"code","source":["!pip install pyBigWig\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"d3c8aI3Wk14K","executionInfo":{"status":"ok","timestamp":1753988674769,"user_tz":420,"elapsed":6723,"user":{"displayName":"Jiarui Cui","userId":"15689554921577522760"}},"outputId":"209f4a00-0c4e-4183-e1bf-80412270f18d"},"execution_count":2,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting pyBigWig\n","  Downloading pyBigWig-0.3.24-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (19 kB)\n","Downloading pyBigWig-0.3.24-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (185 kB)\n","\u001b[?25l   \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m0.0/185.8 kB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K   \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m185.8/185.8 kB\u001b[0m \u001b[31m7.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hInstalling collected packages: pyBigWig\n","Successfully installed pyBigWig-0.3.24\n"]}]},{"cell_type":"code","source":["# âš™ï¸ Setup: Mount Google Drive to save final dataset\n","from google.colab import drive\n","drive.mount('/content/drive')\n","# ğŸ”¢ Expression data\n","import pandas as pd\n","import numpy as np\n","\n","eg_map = pd.read_csv(\"EG.name.txt\", sep=\"\\t\", header=None, names=[\"EID\", \"Tissue\"])\n","expr_df = pd.read_csv(\"57epigenomes.RPKM.pc\", sep=r\"\\s+\", header=None)\n","\n","gene_ids = expr_df.iloc[:, 0].astype(str).str.split(\".\").str[0]\n","expr_values = expr_df.iloc[:, 1:]\n","expr_values.index = gene_ids\n","\n","e116_idx = eg_map[eg_map[\"EID\"] == \"E116\"].index[0]\n","e116_col = pd.to_numeric(expr_values.iloc[:, e116_idx], errors='coerce').dropna()\n","e116_log_expr = np.log2(e116_col + 1)\n","\n","print(\"âœ… Loaded expression values:\", e116_log_expr.shape)\n","\n","# ğŸ§¬ Gene annotation\n","gene_info = pd.read_csv(\n","    \"Ensembl_v65.Gencode_v10.ENSG.gene_info\",\n","    sep=r\"\\s+\",\n","    header=None,\n","    names=[\n","        \"gene_id\", \"chr\", \"start\", \"end\", \"strand\", \"biotype\", \"symbol\", \"description\"\n","    ],\n","    usecols=[0, 1, 2, 3, 4, 6]\n",")\n","\n","gene_info[\"gene_id\"] = gene_info[\"gene_id\"].astype(str).str.split(\".\").str[0]\n","gene_info[\"strand\"] = gene_info[\"strand\"].map({1: \"+\", -1: \"-\"})\n","gene_info[\"tss\"] = np.where(gene_info[\"strand\"] == \"+\", gene_info[\"start\"], gene_info[\"end\"])\n","\n","valid_chr = [str(i) for i in range(1, 23)] + [\"X\", \"Y\"]\n","gene_info = gene_info[gene_info[\"chr\"].isin(valid_chr)]\n","gene_info = gene_info[gene_info[\"gene_id\"].isin(e116_log_expr.index)]\n","print(f\"âœ… Total matched genes: {len(gene_info)}\")\n","\n","# ğŸ“ˆ Histone signals extraction\n","import pyBigWig\n","from tqdm import tqdm\n","\n","bw_k4me3 = pyBigWig.open(\"E116-H3K4me3.fc.signal.bigwig\")\n","bw_k27ac = pyBigWig.open(\"E116-H3K27ac.fc.signal.bigwig\")\n","bw_k36me3 = pyBigWig.open(\"E116-H3K36me3.fc.signal.bigwig\")\n","bw_list = [bw_k4me3, bw_k27ac, bw_k36me3]\n","\n","def extract_histone_signal(chrom, tss, bins=100, flank=5000):\n","    bin_size = 2 * flank // bins\n","    starts = [tss - flank + i * bin_size for i in range(bins)]\n","    ends = [s + bin_size for s in starts]\n","\n","    features = []\n","    for bw in bw_list:\n","        row = []\n","        for start, end in zip(starts, ends):\n","            try:\n","                vals = bw.values(f\"chr{chrom}\", max(0, start), end, numpy=True)\n","                row.append(np.nanmean(vals))\n","            except:\n","                row.append(0.0)\n","        features.append(row)\n","    return np.array(features).T  # shape [100, 3]\n","\n","# ğŸ” Extract across all genes\n","X_all, y_all, gene_ids_all = [], [], []\n","print(\"â³ Extracting histone features...\")\n","\n","for _, row in tqdm(gene_info.iterrows(), total=len(gene_info)):\n","    gene_id = row[\"gene_id\"]\n","    chrom = row[\"chr\"]\n","    tss = int(row[\"tss\"])\n","\n","    try:\n","        signal = extract_histone_signal(chrom, tss)\n","        if signal.shape == (100, 3):\n","            X_all.append(signal)\n","            y_all.append(e116_log_expr[gene_id])\n","            gene_ids_all.append(gene_id)\n","    except:\n","        continue\n","\n","X = np.stack(X_all)\n","y = np.array(y_all)\n","gene_ids = np.array(gene_ids_all)\n","\n","print(f\"âœ… Final shapes: X={X.shape}, y={y.shape}\")\n","\n","# ğŸ’¾ Save to Google Drive /content/drive/MyDrive/Gene_Expression_Prediction_Histone_Marks/Phase_1/Dataset/expression_chr22.npz\n","np.savez(\"/content/drive/MyDrive/Gene_Expression_Prediction_Histone_Marks/Phase_1/Dataset/expression_E116_all.npz\", X=X, y=y, gene_ids=gene_ids)\n","print(\"ğŸ“ Saved to /content/drive/MyDrive/Gene_Expression_Prediction_Histone_Marks/Phase_1/Dataset/expression_E116_all.npz\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"qtB4fu2C5dV2","executionInfo":{"status":"ok","timestamp":1753989601296,"user_tz":420,"elapsed":410044,"user":{"displayName":"Jiarui Cui","userId":"15689554921577522760"}},"outputId":"ee4db2de-e6b8-4a91-bdd0-810eb289e986"},"execution_count":8,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]},{"output_type":"stream","name":"stderr","text":["/tmp/ipython-input-1939934608.py:9: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57) have mixed types. Specify dtype option on import or set low_memory=False.\n","  expr_df = pd.read_csv(\"57epigenomes.RPKM.pc\", sep=r\"\\s+\", header=None)\n"]},{"output_type":"stream","name":"stdout","text":["âœ… Loaded expression values: (19795,)\n","âœ… Total matched genes: 19802\n","â³ Extracting histone features...\n"]},{"output_type":"stream","name":"stderr","text":[" 97%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–‹| 19186/19802 [06:19<00:11, 52.49it/s]/tmp/ipython-input-1939934608.py:61: RuntimeWarning: Mean of empty slice\n","  row.append(np.nanmean(vals))\n","100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 19802/19802 [06:31<00:00, 50.63it/s]\n"]},{"output_type":"stream","name":"stdout","text":["âœ… Final shapes: X=(19802, 100, 3), y=(19802,)\n","ğŸ“ Saved to /content/drive/MyDrive/Gene_Expression_Prediction_Histone_Marks/Phase_1/Dataset/expression_E116_all.npz\n"]}]}]}